packages/chain-adapters/src/solana/SolanaChainAdapter.ts

```diff
@@ -0,0 +1,354 @@
+import { microLamportsToLamports } from './utils'
+
+export interface ChainAdapterArgs {
+  providers: {
+    http: unchained.solana.Api
+    ws: unchained.ws.Client<unchained.solana.Tx>
+  }
+  rpcUrl: string
+}
+
+export class ChainAdapter implements IChainAdapter<KnownChainIds.SolanaMainnet> {
+  static readonly defaultBIP44Params: BIP44Params = {
+    purpose: 44,
+    coinType: Number(ASSET_REFERENCE.Solana),
+    accountNumber: 0,
+  }
+
+  protected readonly chainId = solanaChainId
+  protected readonly assetId = solAssetId
+
+  protected readonly providers: {
+    http: unchained.solana.Api
+    ws: unchained.ws.Client<unchained.solana.Tx>
+  }
+
+  protected connection: Connection
+  protected parser: unchained.solana.TransactionParser<unchained.solana.Tx>
+
+  constructor(args: ChainAdapterArgs) {
+    this.providers = args.providers
+
+    this.connection = new Connection(args.rpcUrl)
+
+    this.parser = new unchained.solana.TransactionParser({
+      assetId: this.assetId,
+      chainId: this.chainId,
+    })
+  }
+
+  getName() {
+    const enumIndex = Object.values(ChainAdapterDisplayName).indexOf(ChainAdapterDisplayName.Solana)
+    return Object.keys(ChainAdapterDisplayName)[enumIndex]
+  }
+
+  getDisplayName() {
+    return ChainAdapterDisplayName.Solana
+  }
+
+  getType(): KnownChainIds.SolanaMainnet {
+    return KnownChainIds.SolanaMainnet
+  }
+
+  getFeeAssetId(): AssetId {
+    return this.assetId
+  }
+
+  getChainId(): ChainId {
+    return this.chainId
+  }
+
+  getBIP44Params({ accountNumber }: GetBIP44ParamsInput): BIP44Params {
+    if (accountNumber < 0) {
+      throw new Error('accountNumber must be >= 0')
+    }
+    return { ...ChainAdapter.defaultBIP44Params, accountNumber }
+  }
+
+  async getAddress(input: GetAddressInput): Promise<string> {
+    try {
+      const { accountNumber, pubKey, wallet, showOnDevice = false } = input
+
+      if (pubKey) return pubKey
+
+      if (!supportsSolana(wallet)) throw new Error('Wallet does not support Solana.')
+
+      const address = await wallet.solanaGetAddress({
+        addressNList: toAddressNList(this.getBIP44Params({ accountNumber })),
+        showDisplay: showOnDevice,
+      })
+
+      if (!address) throw new Error('Unable to generate Solana address.')
+
+      return address
+    } catch (err) {
+      return ErrorHandler(err)
+    }
+  }
+
+  async getAccount(pubkey: string): Promise<Account<KnownChainIds.SolanaMainnet>> {
+    try {
+      const data = await this.providers.http.getAccount({ pubkey })
+
+      const balance = BigInt(data.balance) + BigInt(data.unconfirmedBalance)
+
+      return {
+        balance: balance.toString(),
+        chainId: this.chainId,
+        assetId: this.assetId,
+        chain: this.getType(),
+        chainSpecific: {
+          tokens: data.tokens.map(token => ({
+            assetId: toAssetId({
+              chainId: this.chainId,
+              assetNamespace: 'spl',
```

Comment:
I'm not familiar with Solana's CAIP spec, but I couldn't find reference to `spl` - is this correct?

Reply:
To explain the choice: I couldn't find way to describe the SPL token namespace in the CAIP spec too

So, I followed the same logic from Ethereum or other chains where they namespace tokens by their spec name (ERC20, BEP20, BIP), SPL is the commonly used name for the System Program Library where SPL Tokens are non-native tokens

To be more descriptive, we would need to add `token` to the namespace as SPL is larger than we think, but as this namespace is the `assetNamespace`, I think we shouldn't be too verbose

Reply:
FYI @NeOMakinG, there *is* a CAIP spec for Solana assets, with `token` and `nft` `asset_namespace`s: https://namespaces.chainagnostic.org/solana/caip19
We should strive to be spec-compliant whenever possible and divert from it if a spec doesn't exist and contribute back (as we've done for Osmosis assets back then in https://github.com/ChainAgnostic/CAIPs/pull/111).

Reply:
> FYI @NeOMakinG, there _is_ a CAIP spec for Solana assets, with `token` and `nft` `asset_namespace`s: https://namespaces.chainagnostic.org/solana/caip19 We should strive to be spec-compliant whenever possible and divert from it if a spec doesn't exist and contribute back (as we've done for Osmosis assets back then in [ChainAgnostic/CAIPs#111](https://github.com/ChainAgnostic/CAIPs/pull/111)).

Agree, I didn't find it at the time but seems the right way for sure, we need to update it from the `ASSET_NAMESPACE` constant, asset generation and mocks!

Reply:
Changed it in https://github.com/shapeshift/web/pull/7930

Reply:
Going to be merged very soon, will require a rebase!

Reply:
ffa60130e017037b980f974baf64ce4eb622435d